### DISCUSSION TITLE
using bashly without ruby or docker

### DISCUSSION THREAD

**Discussion by elee1766** [bashly-framework/bashly#541] [Ideas]
### Description

hi.

i really love bashly, it's honestly life changing, but i really hate that i either need a ruby install, or docker, to use it. it's the only thing i have ruby installed on my system for, since i dont like waiting for docker every time to build. (i guess i could create a daemon using compose that mounted my code dir and run the bashly commands there.... its lots of effort regardless)

i am not very familiar with ruby packaging and bundling, but I thought it might be a fun side-project to try to compile a giant redistributable bashly binary.

my thought was maybe i could try to bundle bashly and then just put that code into some binary with a ruby runtime and package that together.

i was curious if you had any experience with this, or if you knew of any weird libraries or ruby functionality that you are using within bashly which might make this difficult or impossible? or maybe i am on a wrong track and there is a hidden secret easy way to use bashly as single binary

if not i will try to do some research and see if i can hack something out.

**Comment by DannyBen** [comment]
I hear you. I also wish Ruby apps could be distributed as a single executable.
I looked for some solutions in the past, none came close to attaining this goal gracefully enough.

Also, if I recall correctly, one of the issues to consider are dependencies that need to be compiled - I think bashly has at least the YAML library (Psych) that is system-specific.

I am curious to see if you find something new in this regard, feel free to share your progress - happy to help.

**Comment by elee1766** [comment]
seems this is a maintained tool for it https://github.com/tamatebako/tebako

i will play around with it 

https://github.com/tamatebako/tebako/blob/6ccc63292ba92b7c51b99b2e7d79627eecca1ac2/tests-2/fixtures/gems-psych-5/Gemfile#L3

they seem to have an example with psych 5 that works? but im not sure about other packages.

naive compilation with `tebako press --entry-point ./bin/bashly --root . --Ruby=3.2.4` failed with 

```
   ... @ /home/a/.tebako/o/s/bin/bundle exec /home/a/.tebako/o/s/bin/gem build /home/a/repo/github.com/DannyBen/bashly/bashly.gemspec
tebako-packager failed: Failed to run /home/a/.tebako/o/s/bin/bundle exec /home/a/.tebako/o/s/bin/gem build /home/a/repo/github.com/DannyBen/bashly/bashly.gemspec (pid 399942 exit 1):
 Ignoring debug-1.9.2 because its extensions are not built. Try: gem pristine debug --version 1.9.2
Ignoring io-console-0.7.2 because its extensions are not built. Try: gem pristine io-console --version 0.7.2
Ignoring strscan-3.1.0 because its extensions are not built. Try: gem pristine strscan --version 3.1.0
Ignoring debug-1.9.2 because its extensions are not built. Try: gem pristine debug --version 1.9.2
Ignoring io-console-0.7.2 because its extensions are not built. Try: gem pristine io-console --version 0.7.2
Ignoring strscan-3.1.0 because its extensions are not built. Try: gem pristine strscan --version 3.1.0
/home/a/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.5.11/lib/bundler/definition.rb:594:in `materialize': Could not find debug-1.9.2, psych-5.1.2, io-console-0.7.2, stringio-3.1.1, strscan-3.1.0 in locally installed gems (Bundler::GemNotFound)
        from /home/a/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.5.11/lib/bundler/definition.rb:193:in `specs'
        from /home/a/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.5.11/lib/bundler/definition.rb:259:in `specs_for'
        from /home/a/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.5.11/lib/bundler/runtime.rb:18:in `setup'
        from /home/a/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.5.11/lib/bundler.rb:163:in `setup'
        from /home/a/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.5.11/lib/bundler/setup.rb:29:in `block in <top (required)>'
        from /home/a/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.5.11/lib/bundler/ui/shell.rb:159:in `with_level'
        from /home/a/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.5.11/lib/bundler/ui/shell.rb:111:in `silence'
        from /home/a/.tebako/o/s/lib/ruby/gems/3.2.0/gems/bundler-2.5.11/lib/bundler/setup.rb:29:in `<top (required)>'
        from <internal:/home/a/.tebako/o/s/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:86:in `require'
        from <internal:/home/a/.tebako/o/s/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:86:in `require'
        from /home/a/.tebako/o/s/lib/ruby/3.2.0/rubygems.rb:1370:in `<top (required)>'
        from <internal:gem_prelude>:2:in `require'
        from <internal:gem_prelude>:2:in `<internal:gem_prelude>'
```

maybe im just doing something wrong, might ask them for help.

**Comment by DannyBen** [comment]
Seems like they complain about some built-in Ruby gems - like the `debug` gem and `io/console`.
One guess would be that the system this is built on does not have the proper OS dependencies to compile these gems.

In my [Alpine Ruby image](https://github.com/DannyBen/docker-alpine-ruby/blob/master/Dockerfile#L4-L7) for example, I am adding these dependencies:

```dockerfile
ENV BUILD_PACKAGES bash curl curl-dev ruby-dev build-base
ENV RUBY_PACKAGES \
  ruby ruby-io-console ruby-irb \
  ruby-json ruby-etc ruby-bigdecimal ruby-rdoc \
  libffi-dev zlib-dev yaml-dev
```

**Reply by elee1766** [reply]
they have an alpine based docker image. 

if you are already building in alpine, then maybe this is a good start. I'll try to copy any system libs via the alpine build process you have (i downloaded a bunch of debian libs on my personal computer, who knows if i got them all)

**Reply by elee1766** [reply]
compiling with said packages installs nets me the same results. it seems tebako isnt building any gems with extensions. 

maybe i'll try to create a smaller bundle with the same packages from their examples, then add some more you are using, to see if things go wrong there or not

**Reply by DannyBen** [reply]
Yes, the bashly docker is built from that dannyben/alpine-ruby image.
If this tebako cannot handle extensions it would make it quite useless. Many Ruby dependencies, and even ones that are considered "standard library" have native extensions.

But starting small with an example they provide and adding more gems on top of that sounds like a good approach.

**Reply by elee1766** [reply]
yeah, from what i see, people are using sinatra and rails with it - which i assume both have extensions...

**Comment by keilmillerjr** [comment]
Bashly is a development tool, and not intended to be distributed with your code. They are independent of each other. Your code produced will run as an executable on any system with bash, such as Linux and Mac OSX.

Setting up a ruby dev environment is easy. I have used RVM in the past to manage different ruby versions and gem sets per project. I am currently using ASDF, and highly recommend it. It is extremely easy to set up. It is totally optional. Just having ruby and installing the bashly gem is all that is needed. Mac OSX already has ruby pre installed. On arch, installing ruby is a one liner.

```
pacman -S ruby && gem install bashly
```

**Reply by elee1766** [reply]
bashly is a development tool, that I want to vendor with my code. this is a very normal workflow.

my desire is to uninstall Ruby from my system. having it in any of my systems is an eyesore. ideally it can be removed from all systems, and maybe the face of the planet. 

I already have Ruby envs managed with mise. hopefully I can remove all of this if I'm successful.

I think you're misunderstanding what I'm trying to do. I want to not have ruby installed while developing my cli :)

it doesn't matter how easy Ruby is to set up; I know it's one command I have it installed. I just would prefer to not have it installed in any capacity in my system when I'm literally only using it to run bashly to compile my clis. it's polluting my path, taking up disk space, and increasing upgrade times. it's a complete pain.

you can read the original post - it might bring more context to my problem. I would like to bundle without ruby. it's a bit silly that you need ruby to bundle the script but not to run the script.

**Reply by pcrockett** [reply]
FWIW this is one reason a lot of people love Go. It's really nice to be able to deploy a single binary that just works without any ceremony.

I maintain [asdf-bashly](https://github.com/pcrockett/asdf-bashly) and yeah... if a static, dependency-free executable existed, I would use that in a heartbeat. Users wouldn't need to consult the README at all; they could just install the plugin without realizing Ruby is a dependency, reducing cognitive load for everyone.
